### DISCUSSION TITLE
Query at runtime whether Control property can be assigned from URI?

### DISCUSSION THREAD

**Discussion by kuiperzone** [AvaloniaUI/Avalonia#6997] [Q&A] [answered]
Hi,

I am writing a AXAML processor. In (A)XAML, I can write:

            <Image Grid.Column="0" Grid.Row="0"
                Source="/Assets/testcard.png" Stretch="Uniform"/>

Here, I just "know" that "Source" can populated from a path, and I know that the path will be loaded as an image at runtime. But the same does not apply to `TextBlock.Text` from example, which is just a string.

**But if, in code, I was querying every Control property, how could I determine whether the string I want to assign should be treated as a URI or not?** In other words, is there any way to distinguish that Image.Source is a URI and that TextBlock.Text is just text?

Is there some list finite list or properties, some way of querying, or a heuristic I could use?

For example, I could key off the property name such as "Source". But then there is "Window.Icon" which can also be loaded as a URI in the XAML. I would imagine there are others.

Help appreciated on this one

Thanks

**Comment by kekekeks** [comment, marked as answer]
It's controlled by the property type. We allow converting from string if:
1) Type has `System.ComponentModel.TypeConverterAttribute`
2) Type has `T Parse(string)` or `T Parse(string s, CultureInfo info)` public static method
3) (for historical reasons and for some external types) Type converter is specified in [this list](https://github.com/AvaloniaUI/Avalonia/blob/5ece272a597ead5875fed27a7baf1cdc8a05324f/src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/AvaloniaXamlIlLanguage.cs#L117-L126)

In case of IImage its the third one. Currently only BitmapTypeConverter and StyleInclude/ResourceInclude look for resource uris.

**Comment by kuiperzone** [comment]
A property type assignable to one of the following should do it?:

    Type[] s_assetTypes = new[] { typeof(IImage), typeof(IResourceProvider), typeof(ResourceInclude) };
